### ISSUE TITLE
Image messages not wokring on email

### ISSUE THREAD

**Issue by Soullivaneuh** [gitlabhq/gitlabhq#9063]
If a user add an image on an issue, this will be rendered like this on email:

![gitlab_image_email](https://cloud.githubusercontent.com/assets/1698357/6919994/e9a66d38-d7bb-11e4-87e1-37aa88b6d76c.png)

Same result on Android Gmail app.

Please note that base64 inline encode for image does not work on many email manager, web or client.

GitLab 7.9.1 6f6c5f1

**Comment by DouweM** [comment]
I just tested it myself, and I can confirm it's not working on Gmail, on Android or web. I'm pretty sure we tested it in multiple browsers and clients, so I'm not sure how this happened.

Can you comment @123Haynes? If this isn't working on these platforms, we should roll back the feature and settle for simply linking to the images.

**Comment by Soullivaneuh** [comment]
@DouweM @123Haynes 

Base64 image support is actually very bad on most email application and **shouldn't be use**.

Maybe later, but surely not now. :)

Take a look here: http://stackoverflow.com/a/18646173/1731473

**Comment by 123Haynes** [comment]
@Soullivaneuh when we implemented this we tested it with iphone, thunderbird, gmail web client, gmail on android and K-9 Mail on Android.
It's weird that it renders the html tag. That looks more like a bug with html escaping just from the screenshot. 
But I'll take a closer look at it later today.

The post you linked at is targeted at displaying the image even if normal attached images are blocked. But you may be right. We'll see after a short analysis :)

**Comment by Soullivaneuh** [comment]
@123Haynes Note sure of what you saying about "html escaping" on "my app".

It's just a classic gmail android app and classic gmail browser app.

I'm sure I'm not the only one to have this issue. :smiling_imp:

**Comment by Soullivaneuh** [comment]
@123Haynes Can I see in which PR this was introduce?

Thanks.

**Comment by 123Haynes** [comment]
I just managed to reproduce it somewhat. K-9 Mail works fine, Gmail app and mobile website displays only the alt text for me. Not the complete source. Forwarded it to my work email account to test it with outlook 2010 and it displays a placeholder image.
So this definatly doesn't work as expected, but probably isn't as bad as it looked on the screenshot.
I think we should target 7.10 to come up with a working fix.

Only other possiblity I see to get images in emails to work would be to add the images as attachments and then modify the locations similar to how the current filter replaces the location with base64 encoded images.

@Soullivaneuh sure. We introduced it with this MR:
https://gitlab.com/gitlab-org/gitlab-ce/merge_requests/373

**Comment by DouweM** [comment]
@123Haynes I'm seeing the complete source in Gmail app and mobile.

**Comment by 123Haynes** [comment]
@DouweM weird. What do you think about the inline images option? I'm not sure if it's possible to add the attachments from the filter. 

Another possible solution would be to replace all images with the highlighted alt text of the images instead of displaying them.

**Comment by DouweM** [comment]
@123Haynes Personally, I think it's not worth the effort and we should just settle for replacing the images by a link.

**Comment by Soullivaneuh** [comment]
Agree with @DouweM.

Classic image with link works well in any email app, inline encoded doesn't.

I think the compliance element is the more important parameter for this kind of decision.

**Comment by Soullivaneuh** [comment]
I just see the related PR.

> This MR fixes the broken images in emails that occured scince we introduced access control for all attachments in 7.8.

Why not just has a user-data system like GitHub? Attached image on issue generate a strong link with random ID.

Image are public but can't be retrieved without the link.

**Comment by 123Haynes** [comment]
@Soullivaneuh That's what we had until 7.8. 
But this poses a possible data leak which would be very bad for company installations.
That's why we changed it so access for every attachment is validated.

**Comment by Soullivaneuh** [comment]
I don't understand why this system could be unsecured with strong links.

GitHub uses this system even if for private repo. I can't found any data leak possibility, retrieve those links is harder than crack a password.

It's a must have if you want to get all email working.

**Comment by stanhu** [comment]
There is no such thing as a "strong link". This is just security through obscurity. If a company installation were to use HTTP, then those links would be visible by a network snooper.

**Comment by Soullivaneuh** [comment]
> If a company installation were to use HTTP

I'm sorry but I strongly disagree with that. A company who care a lot about data security **does not** use HTTP protocol. It always use HTTPS.

In any way, inline encoded image doesn't work on lot of email application and is not recommended to use it.

So, seems to be obvious that inline images on emails are not a solution.

**Comment by 123Haynes** [comment]
@Soullivaneuh noone argues with that.

Currently we have 3 options:
* allow access to all images so they are just secured by the obsure url (I don't like this one)
* replace image tags with links to the images (probably the easiest solution, but images won't be displayed in emails)
* Use inline attachments (I haven't looked into this to much yet. But you'd attach all images as normal attachments and then replace the links in the text so they point to the attachments)

**Comment by Soullivaneuh** [comment]
My opinion about your 3 options:

> allow access to all images so they are just secured by the obsure url (I don't like this one)

I think is the easy way to do for being the more compliant with all browsers/applications. I like this one and I don't think is unsecured.

> replace image tags with links to the images (probably the easiest solution, but images won't be displayed in emails)

So it's resolving an issue by hiding it. We will totally remove a functionality with that. It's important to have a quick view of a ticket on our phone for example...

> Use inline attachments (I haven't looked into this to much yet. But you'd attach all images as normal attachments and then replace the links in the text so they point to the attachments)

Don't know how attachments works and if show them on messages is compliant with any system. Have to be careful with that.

**Comment by Soullivaneuh** [comment]
I have another more friendly and quick solution to propose.

Let's administrator choice from config file and/or application settings, like this:

```yml
image_acl_protection: true # default
image_base64_encore: true # default
```

So if administrators want more compliant image system and are OK with public link, they will just have to set both params to `false`.

What do you think about it?

**Comment by 123Haynes** [comment]
If possible I'd like a solution that doesn't require a config setting because this introduces unneeded complexity. 
I'll only consider that if we can't come to an agreement.

I'll take a closer look at option 1 and 3 in the comming days.
If I find out that our urls for images are indeed secure enough I don't mind excluding images from the  access control.
But for now I'm still against relying on obscure urls as the only security measure.

**Comment by Soullivaneuh** [comment]
If you swill against after investigation, only solution 3 could be a big deal only if it's compliant to all email application like linked image does.

In other case, only configuration file could make an agreement about that. But I agree, it's not the best solution.

**Comment by stanhu** [comment]
@123Haynes, I tested with Gmail does when you embed an image into a message. They follow this practice:

http://stackoverflow.com/questions/4018709/how-to-create-an-email-with-embedded-images-that-is-compatible-with-the-most-mai

**Comment by 123Haynes** [comment]
@stanhu I think that is what I meant with option 3. You add it as an attachment and modify the image urls to use the cid. 

I did some research regarding the first option. Here are the results:
* The urls can not easily be found out by an attacker **if** HTTPS is being used. 
* They are not recommended as the only security measure to protect sensible information. The general recommendation seems to be to use a password/username and the urls.

In the end it's a tradeoff between security and usability.
The URLs can be accessed by anyone who knows the link. That makes it possible to use them in emails. But at the same time that **can** be a problem because you can't limit the access to certain users.

An example (This probably won't happen often, but it's enough to get the idea):
Imagine an employee leaves the company on bad terms and his account gets blocked. 
He can't access gitlab anymore now, but he still has a link to a screenshot that contains sensible information.
After a few weeks he decides to post that link online. Result => possible image loss, data leak and possible attacks with the information from the screenshot.
The company has to manually access the server and delete that image now, because they can't control who accesses it. (Keep in mind that you can't delete uploaded attachments from the gitlab gui at the moment)

As I said before: this is very unlikely and is a worst case scenario.
Facebook decided to use the urls in favour of usability for example, but for them sharing pictures is a feature. 
They wan't all people who know the link to be able to access a picture.
But usually these pics don't contain very sensitve informations and you can delete them easily.


I think the best solution would be the inline attachments if they work in most email programs.
I'm not sure how that would work in ruby yet. Probably something like:
1. Add the attachments with the Actionmailer
2. Invoke a html-pipeline filter which gets the attachments list
3. replace the image src with a link to the image cid

I have a training in the beginning of may to learn a bit more about webapplication security.
I'll ask the pentesters who'll be present there how they would handle this situation.(I want to fix this sooner if possible, just letting you know that we might need to revisit whatever solution we come up with later.)
Regardless of what we decide, I think we can all agree that the current behaviour needs to be reverted if it shows the whole source of the image tag for some clients. (I'll make a MR for that soon)

/cc @DouweM @Soullivaneuh

**Comment by Soullivaneuh** [comment]
> 3. replace the image src with a link to the image cid

I'm not sure to understand. Will email got image on body or just a link to see it? If is just a link, this is like removing completely the image feature.

**Comment by stanhu** [comment]
Right, I agree with the MR revert. I also agree that an open but obscure URL should NOT be done here; there's no real security there.

The link to the image (and requiring a login to access) would be more secure simply because you're not duplicating potentially sensitive information into people's e-mails. That being said, I think the decision to use inline attachments or including links really depends on the organization's security policies. I can see a case where one organization wants the convenience of inline attachments, but other organizations would prefer to disable that option. Ideally, there would be a configuration option to select between the two.

**Comment by Soullivaneuh** [comment]
> Ideally, there would be a configuration option to select between the two.

I agree with that and it's globally what I said here: https://github.com/gitlabhq/gitlabhq/issues/9063#issuecomment-88595946

It's the best solution unless we find a way to have secure image by username/password that **can be shown** on email without base64 encore inline method, that not working at all on lot of popular email application.

**Comment by 123Haynes** [comment]
So I've talked a bit with @DouweM about this.
We will probably go with the urls for 7.10 simply because it requires only minimal changes and we think that images in emails are important. We realize that this will be a little backstep regarding the security from 7.9.
This doesn't mean that we will keep using that solution after 7.10.
If we can improve the situation later, we will do that.
Inline attachments might be an option if they work on all common mail clients(That needs to be throughoutly tested first)

PS: We won't implement an option to change between the 2 methods. That would just add an unnecessary switch because in the end they have the same effect for the end user.

@Soullivaneuh with the inline attachments the pictures would still be shown in the email, but the src attribute would link to the attached images instead of the ones on the server.

**Comment by DouweM** [comment]
It is technically "security through obscurity" to leave the URLs publicly accessible behind a 32-char hex and file name, but I think that's sufficient in this case.

**Comment by Soullivaneuh** [comment]
Ok for me.

Attachements solution seems to be a good deal too. :+1:

**Comment by Soullivaneuh** [comment]
Any new for a fix? I see that 7.10 is on RC. Would be very great to have this fix on this release.

Thanks.

**Comment by DouweM** [comment]
@Soullivaneuh We're working on it, it will hopefully be ready for a later RC.

**Comment by Soullivaneuh** [comment]
I'm glad to hear it, thanks! :+1:

**Comment by 123Haynes** [comment]
fixed in 7.10

**Comment by qaisjp** [comment]
Which commit??
